src/auth/mod.rs

```diff
@@ -0,0 +1,178 @@
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types for parsing and authenticating HAWK headers.
+//! Matches the [Python logic](https://github.com/mozilla-services/tokenlib).
+//! We may want to extract this to its own repo/crate in due course.
+
+#[cfg(test)]
+mod test;
+
+use std::{
+    error::Error,
+    fmt::{self, Display, Formatter},
+};
+
+use actix_web::{error::ResponseError, http::header::ToStrError, FromRequest, HttpRequest};
+use base64::{self, DecodeError};
+use chrono::offset::Utc;
+use hawk::{Error as HawkError, Header as HawkHeader, Key, RequestBuilder};
+use hkdf::{Hkdf, InvalidLength};
+use hmac::{
+    crypto_mac::{InvalidKeyLength, MacError},
+    Hmac, Mac,
+};
+use ring;
+use serde_json::{self, Error as JsonError};
+use sha2::Sha256;
+use time::Duration;
+
+use settings::Settings;
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+pub struct HawkPayload {
+    pub expires: f64,
+    pub node: String,
+    pub salt: String,
+    pub uid: u64,
+}
+
+impl HawkPayload {
+    fn new(
+        header: &str,
+        method: &str,
+        path: &str,
+        host: &str,
+        port: u16,
+        settings: &Settings,
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        if &header[0..5] != "Hawk " {
+            return Err(AuthError);
+        }
+
+        let header: HawkHeader = header[5..].parse()?;
+        let id = header.id.as_ref().ok_or(AuthError)?;
+
+        let payload = HawkPayload::extract_and_validate(id, &settings.master_token_secret, expiry)?;
+
+        let token_secret = hkdf_expand_32(
+            format!("services.mozilla.com/tokenlib/v1/derive/{}", id).as_bytes(),
+            Some(payload.salt.as_bytes()),
+            &settings.master_token_secret,
+        )?;
+        let token_secret = base64::encode_config(&token_secret, base64::URL_SAFE);
+
+        let request = RequestBuilder::new(method, host, port, path).request();
+        if request.validate_header(
+            &header,
+            &Key::new(token_secret.as_bytes(), &ring::digest::SHA256),
+            // Allow plenty of leeway for clock skew, because
+            // client timestamps tend to be all over the shop
+            Duration::weeks(52),
+        ) {
+            Ok(payload)
+        } else {
+            Err(AuthError)
+        }
+    }
+
+    fn extract_and_validate(
+        id: &str,
+        master_secret: &[u8],
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        let decoded_id = base64::decode_config(id, base64::URL_SAFE)?;
+        if decoded_id.len() <= 32 {
+            return Err(AuthError);
+        }
+
+        let payload_length = decoded_id.len() - 32;
+        let payload = &decoded_id[0..payload_length];
+        let signature = &decoded_id[payload_length..];
+
+        let signing_secret = hkdf_expand_32(
+            b"services.mozilla.com/tokenlib/v1/signing",
+            None,
+            master_secret,
+        )?;
```

Comment:
In a future refactor, you could consider pre-calculating this and caching it somewhere globally.  The overhead will be pretty minimal in practice though so NBD.

Reply:
We're going to need to implement a custom `Deserialize` for this property on `Settings` anyway, so could do it in there.

Reply:
Ref: #26